system/lib/libc/crt1.c

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2019 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ */
+
+// Much of this code comes from:
+// https://github.com/CraneStation/wasi-libc/blob/master/libc-bottom-half/crt/crt1.c
+
+#include <stdlib.h>
+#include <sysexits.h>
+#include <wasi/wasi.h>
+#include <stdio.h>
+
+extern void __wasm_call_ctors(void) __attribute__((weak));
+extern int main(int argc, char** argv);
+
+static __wasi_errno_t populate_args(size_t *argc, char ***argv) {
+  __wasi_errno_t err;
+
+  /* Get the sizes of the arrays we'll have to create to copy in the args. */
+  size_t argv_buf_size;
+  err = __wasi_args_sizes_get(argc, &argv_buf_size);
+  if (err != __WASI_ESUCCESS) {
+    return err;
+  }
+  if (*argc == 0) {
+    return __WASI_ESUCCESS;
+  }
+
+  /* Allocate memory for the array of pointers, adding null terminator. */
+  *argv = malloc(sizeof(char *) * (*argc + 1));
```

Comment:
I guess this means it's impossible to compile a program that doesn't include `malloc`. Is that possible on other platforms?

Reply:
Interesting point.    This is how wasi works today :(

I wonder if we should bring that up in wasi subgroup?

Reply:
Actually looking at this code it should be possible to use calloc rather than malloc and put this stuff on the stack before calling main.  That would be more inline with other platforms too.